### PR TITLE
[BEAM-2825] Refactored SparkGroupAlsoByWindowViaWindowSet to improve …

### PR DESCRIPTION
**Pull request by staslev**
…readability.

Follow this checklist to help us incorporate your contribution quickly and easily:

 - [x] Make sure there is a [JIRA issue](https://issues.apache.org/jira/projects/BEAM/issues/) filed for the change (usually before you start working on it).  Trivial changes like typos do not require a JIRA issue.  Your pull request should address just this issue, without pulling in other changes.
 - [x] Each commit in the pull request should have a meaningful subject line and body.
 - [x] Format the pull request title like `[BEAM-XXX] Fixes bug in ApproximateQuantiles`, where you replace `BEAM-XXX` with the appropriate JIRA issue.
 - [x] Write a pull request description that is detailed enough to understand what the pull request does, how, and why.
 - [ ] Run `mvn clean verify` to make sure basic checks pass. A more thorough check will be performed on your pull request automatically.
 - [ ] If this contribution is large, please file an Apache [Individual Contributor License Agreement](https://www.apache.org/licenses/icla.pdf).

---

This commit is mostly about moving things around, moving anonymous classes to upper levels, extracting methods, minor renaming, formatting, etc.. 

This PR will probably need to be rebased off #3749.

### CODE REVIEW

**Comment by staslev** [comment]
retest this please

**Comment by staslev** [comment]
Run Spark ValidatesRunner

**Comment by staslev** [comment]
R: @amitsela

**Comment by staslev** [comment]
Run Spark ValidatesRunner

File: runners/spark/src/main/java/org/apache/beam/runners/spark/stateful/SparkGroupAlsoByWindowViaWindowSet.java [hunk 1]
```diff
@@ -73,440 +73,555 @@
...
+      // possible input scenarios for (K, Seq, Option<S>):
+      // (1) Option<S>.isEmpty: new data with no previous state.
+      // (2) Seq.isEmpty: no new data, but evaluating previous state (timer-like behaviour).
+      // (3) Seq.nonEmpty && Option<S>.isDefined: new data with previous state.
+
+      final SystemReduceFn<K, InputT, Iterable<InputT>, Iterable<InputT>, W> reduceFn =
+          SystemReduceFn.buffering(wvCoder.getValueCoder());
+
+      // use in memory Aggregators since Spark Accumulators are not resilient
+      // in stateful operators, once done with this partition.
```
Comment (amitsela) [inline runners/spark/src/main/java/org/apache/beam/runners/spark/stateful/SparkGroupAlsoByWindowViaWindowSet.java]: `once done with this partition` maybe I was trying to say `log once done with this partition`?

**Comment by amitsela** [comment]
Only comment I probably have is that I would prefer the "core" functionality on top, and helper methods and classes at the bottom, so that reading through the file starts from "what it does".
So I guess I'd put `groupAlsoByWindow` first, followed by order of usage (the first helper called comes after and so on..).
But that might be just me 😄 

Think abut it, but other than that LGTM.

**Comment by staslev** [comment]
I often hear people say they want the public methods ("core") at the top so that they are easily reachable -  I think that's exactly why they should be at the bottom :)

The bottom of the page stays bottom even if the class evolves and further members (variables, classes) get added at the top (which they are, since we _all_ agree variables should be placed at the top).
If public methods go in the top, the newly added member variables/classes will bump them down. 
This results in the public stuff we want to be easily reachable be now in the **middle** (and we cannot simply skip all the variables), which is sad for all of us.

Makes sense?

File: runners/spark/src/main/java/org/apache/beam/runners/spark/stateful/SparkGroupAlsoByWindowViaWindowSet.java (hunk 1 above)
Comment (staslev) [reply runners/spark/src/main/java/org/apache/beam/runners/spark/stateful/SparkGroupAlsoByWindowViaWindowSet.java]: Yeah I think this comment can be removed, as aggregators are nowhere to be found, probably replaced by `MetricsContainer`s.

**Comment by staslev** [comment]
Run Spark ValidatesRunner

**Comment by staslev** [comment]
@kobisalant, would you mind taking a look to confirm I didn't miss anything when merging your changes into this PR?

**Comment by kobisalant** [comment]
LGTM

**Pull request closed without merging**